odoo/addons/base/models/ir_model.py

```diff
@@ -1689,7 +1689,7 @@ def _reflect_constraint(self, model, conname, type, definition, module, message=
                             write_uid=%s, type=%s, definition=%s, message=%s
                         WHERE id=%s"""
             cr.execute(query, (self.env.uid, type, definition, Json({'en_US': message}), cons_id))
-        return self.browse(cons_id)
+            return self.browse(cons_id)
```

Comment:
Why?  That change does not look right to me.

Reply:
So to give the complete explanation:

- Splitting invalidation between template and the default cache caused some issue because writing on templates didn't invalidate the ormcache of xmlids while _update_xmlids was called in the same process. 
- Adding a clear_cache in _update_xmlids  caused another issue, an increased query count in test_related_fields. The cache was cleared when it wasn't before.
- The call to `_update_xmlids` is actually not useful since nothing changed. We may want to avoid that. Thanks to @odony who noticed this. `_reflect_constraint` return value is only used in  _reflect_model and only to define the new record linked to the constraint. We only need to call `_update_xmlids` if the constraint was create. 

Note: in this case we actually don't need to clear the ormcache but this is a problem for later, we should actually create the ir.model.data but we have no guarantee that it didn't exist, if the record was removed and the ir.model.data not. This is a future improvement, avoid to invalidate the cache if the record was created and not modified. We assume that if the record exist, the ir.model.data is correct.